pkg/releaselock/controller.go

```diff
@@ -94,114 +117,234 @@ func NewLockReleaseController(client kubernetes.Interface, config *LockReleaseCo
+	node := obj.(*corev1.Node)
+	start := time.Now()
+	cmName := ConfigMapNamePrefix + node.Name
+	cm, err := c.client.CoreV1().ConfigMaps(util.ManagedFilestoreCSINamespace).Get(ctx, cmName, metav1.GetOptions{})
+	duration := time.Since(start)
+	c.RecordKubeAPIMetrics(err, metrics.ConfigMapResourceType, metrics.GetOpType, metrics.ReconcilerOpSource, duration)
+
 	if err != nil {
-		klog.Errorf("Failed to get GKE node name from configmap %s/%s: %v", cm.Namespace, cm.Name, err)
-		return err
+		if apiError.IsNotFound(err) {
+			return nil
+		}
+		return fmt.Errorf("failed to get configmap in namespace %s: %v", util.ManagedFilestoreCSINamespace, err)
 	}
-
-	node := nodes[nodeName]
+	klog.Infof("Got configmap (%v) in namespace %s", cm, util.ManagedFilestoreCSINamespace)
 	data := cm.DeepCopy().Data
+
+	var configMapReconcileErrors []error
 	for key, filestoreIP := range data {
-		_, _, _, _, gceInstanceID, gkeNodeInternalIP, err := ParseConfigMapKey(key)
+		err = c.processConfigMapEntryOnNodeCreation(ctx, key, filestoreIP, node, cm)
 		if err != nil {
-			klog.Errorf("Failed to parse configmap key %s: %v", key, err)
-			continue
+			configMapReconcileErrors = append(configMapReconcileErrors, err)
 		}
-		klog.V(6).Infof("Verifying GKE node %s with nodeId %s nodeInternalIP %s exists or not", nodeName, gceInstanceID, gkeNodeInternalIP)
-		nodeExists, err := c.verifyNodeExists(node, gceInstanceID, gkeNodeInternalIP)
-		if err != nil {
-			klog.Errorf("Failed to verify GKE node %s with nodeId %s nodeInternalIP %s still exists: %v", nodeName, gceInstanceID, gkeNodeInternalIP, err)
-			continue
+	}
+	if len(configMapReconcileErrors) > 0 {
+		return errors.Join(configMapReconcileErrors...)
+	}
+	return nil
+
+}
+
+func (c *LockReleaseController) processConfigMapEntryOnNodeCreation(ctx context.Context, key string, filestoreIP string, node *corev1.Node, cm *corev1.ConfigMap) error {
+	_, _, _, _, gceInstanceID, gkeNodeInternalIP, err := ParseConfigMapKey(key)
+	if err != nil {
+		return fmt.Errorf("failed to parse configmap key %s: %v", key, err)
+	}
+	klog.V(6).Infof("Verifying GKE node %s with nodeId %s nodeInternalIP %s exists or not", node.Name, gceInstanceID, gkeNodeInternalIP)
+	entryMatchesNode, err := c.verifyConfigMapEntry(node, gceInstanceID, gkeNodeInternalIP)
+	if err != nil {
+		return fmt.Errorf("failed to verify GKE node %s with nodeId %s nodeInternalIP %s still exists: %v", node.Name, gceInstanceID, gkeNodeInternalIP, err)
+	}
+	if entryMatchesNode {
+		klog.V(6).Infof("GKE node %s with nodeId %s nodeInternalIP %s still exists in API server, skip lock info reconciliation", node.Name, gceInstanceID, gkeNodeInternalIP)
+		return nil
+	}
+
+	// Try to match the latest node, to prevent incorrect releasing the lock in case of a lagging informer/watch
+	latestNode, err := c.client.CoreV1().Nodes().Get(ctx, node.Name, metav1.GetOptions{})
+	if err != nil {
+		if apiError.IsNotFound(err) {
```

Comment:
shall we release the lock if node no longer exists

Reply:
Node deletion should already handle that scenario. Can confirm that lock is released when a node becomes not ready, during my testings.

Reply:
This error handling is merely to distinguish notFound error and other apiErrors. But IMO node not found error is highly unlikely here.

Reply:
so is it something like a node is created and deleted fast when we can hit the not found error?


Can you please elaborate on "Can confirm that lock is released when a node becomes not ready, during my testings." - how is the lock released in that case?

Reply:
In the lock release [v2 design doc](https://docs.google.com/document/d/1ZT4L3JDCenxO8c-vH94Z_4nKhCrpxuqYoTdcVCyJBLM/edit?resourcekey=0-ExA4eJ58o0pmkoV-hW7-eQ&tab=t.0#heading=h.rgrs0o200yll), it is mentioned no-op for node deletion because nfs locks are released gracefully during deletion. v1 design doc also proved that locks are gracefully released during node deletion: [bookmark](https://docs.google.com/document/d/1TDpbkZYmdhQUQld016ynaQ3eX1e84yUEg5oprjUzGog/edit?resourcekey=0-4z_PW8Zf85j5OcrW-iaptg&tab=t.0#bookmark=kix.5n19cf7v29r1)

During my testing, I acquired a file lock on a pod->manually deleted the underlying gce instance of that node->the pod got rescheduled->I confirmed that the file lock could be acquired again, and the entry in the original config map had been removed.

In node.go, during nodeUnstageVolumeUpdateLockInfo, the configmap entry is deleted [here](https://github.com/kubernetes-sigs/gcp-filestore-csi-driver/blob/master/pkg/csi_driver/node.go#L577).

Reply:
Added release lock and delete config map entry to address this extreme corner case